mono/metadata/w32file.c

```diff
@@ -374,40 +374,46 @@ get_filesystem_entries (const gunichar2 *path,
 }
 
 
-MonoArray *
-ves_icall_System_IO_MonoIO_GetFileSystemEntries (MonoString *path,
-						 MonoString *path_with_pattern,
+MonoArrayHandle
+ves_icall_System_IO_MonoIO_GetFileSystemEntries (MonoStringHandle path,
+						 MonoStringHandle path_with_pattern,
 						 gint attrs, gint mask,
 						 gint32 *ioerror)
 {
 	MonoError error;
 	MonoDomain *domain = mono_domain_get ();
-	MonoArray *result;
+	MonoArrayHandle result = MONO_HANDLE_CAST (MonoArray, NULL_HANDLE);
 	int i;
 	GPtrArray *names;
 	
 	*ioerror = ERROR_SUCCESS;
 
-	names = get_filesystem_entries (mono_string_chars (path), mono_string_chars (path_with_pattern), attrs, mask, ioerror);
+	guint32 path_handle = 0;
+	guint32 path_with_pattern_handle = 0;
+	const gunichar2 *const path_chars = mono_string_handle_pin_chars (path, &path_handle);
+	const gunichar2 *const path_with_pattern_chars = mono_string_handle_pin_chars (path_with_pattern, &path_with_pattern_handle);
+	names = get_filesystem_entries (path_chars, path_with_pattern_chars, attrs, mask, ioerror);
+	mono_gchandle_free (path_with_pattern_handle);
+	mono_gchandle_free (path_handle);
 
 	if (!names) {
 		// If there's no array and no error, then return an empty array.
 		if (*ioerror == ERROR_SUCCESS) {
-			MonoArray *arr = mono_array_new_checked (domain, mono_defaults.string_class, 0, &error);
+			MonoArrayHandle arr = mono_array_new_handle (domain, mono_defaults.string_class, 0, &error);
 			mono_error_set_pending_exception (&error);
 			return arr;
 		}
-		return NULL;
+		return result;
 	}
 
-	result = mono_array_new_checked (domain, mono_defaults.string_class, names->len, &error);
+	result = mono_array_new_handle (domain, mono_defaults.string_class, names->len, &error);
 	if (mono_error_set_pending_exception (&error))
```

Comment:
`HANDLES()` functions don't need to call `mono_error_set_pending_exception` then can just do `return_val_if_nok (error, <return value>)`